File: src/main/java/xyz/phanta/cbtweaker/buffer/BufferGroup.java
package xyz.phanta.cbtweaker.buffer;

import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import xyz.phanta.cbtweaker.integration.jei.JeiBufferGroup;

import javax.annotation.Nullable;
import java.util.*;

public class BufferGroup {

    private final SortedMap<BufferType<?, ?, ?, ?>, List<?>> bufTable
            = new TreeMap<>(Comparator.comparing(BufferType::getId));

    @SuppressWarnings("unchecked")
    @Nullable
    public <B> List<B> getBuffers(BufferType<B, ?, ?, ?> bufType) {
        List<B> bufs = (List<B>)bufTable.get(bufType);
        return bufs != null ? bufs : Collections.emptyList();
    }

    @SuppressWarnings("unchecked")
    public <B, A> A accumulateIngredients(BufferType<B, A, ?, ?> bufType) {
        A acc = bufType.createAccumulator();
        List<B> bufs = (List<B>)bufTable.get(bufType);
        if (bufs != null) {
            for (B buf : bufs) {
                bufType.accumulate(acc, buf);
            }
        }
        return acc;
    }

    @SuppressWarnings("unchecked")
    public <B> void addBuffer(BufferType<B, ?, ?, ?> bufType, B buffer) {
        ((List<B>)bufTable.computeIfAbsent(bufType, k -> new ArrayList<>())).add(buffer);
    }

    public void forEach(Visitor visitor) {
        for (Map.Entry<BufferType<?, ?, ?, ?>, List<?>> entry : bufTable.entrySet()) {
            visitUnchecked(visitor, entry.getKey(), entry.getValue());
        }
    }

    @SuppressWarnings("unchecked")
    private static <B> void visitUnchecked(Visitor visitor, BufferType<B, ?, ?, ?> bufType, List<?> buffers) {
        visitor.visit(bufType, (List<B>)buffers);
    }

    public interface Visitor {

        <B> void visit(BufferType<B, ?, ?, ?> bufType, List<B> buffers);

    }

    public static class Factory {

        private final Map<BufferType<?, ?, ?, ?>, List<? extends BufferFactory<?, ?>>> factoryTable = new HashMap<>();

        @SuppressWarnings("unchecked")
        public <B> void addFactory(BufferType<B, ?, ?, ?> bufType, BufferFactory<B, ?> factory) {
            ((List<BufferFactory<B, ?>>)factoryTable.computeIfAbsent(bufType, k -> new ArrayList<>())).add(factory);
        }

        public BufferGroup createBufferGroup(World world, BlockPos pos, @Nullable BufferObserver observer) {
            BufferGroup bufGroup = new BufferGroup();
            for (Map.Entry<BufferType<?, ?, ?, ?>, List<? extends BufferFactory<?, ?>>> entry
                    : factoryTable.entrySet()) {
                createBuffersUnchecked(bufGroup, entry.getKey(), entry.getValue(), world, pos, observer);
            }
            return bufGroup;
        }

        @SuppressWarnings("unchecked")
        private static <B> void createBuffersUnchecked(BufferGroup bufGroup, BufferType<B, ?, ?, ?> bufType,
                                                       List<? extends BufferFactory<?, ?>> factories,
                                                       World world, BlockPos pos, @Nullable BufferObserver observer) {
            for (BufferFactory<B, ?> factory : (List<BufferFactory<B, ?>>)factories) {
                bufGroup.addBuffer(bufType, factory.createBuffer(world, pos, observer));
            }
        }

        public JeiBufferGroup createJeiBufferGroup() {
            JeiBufferGroup bufGroup = new JeiBufferGroup();
            for (Map.Entry<BufferType<?, ?, ?, ?>, List<? extends BufferFactory<?, ?>>> entry
                    : factoryTable.entrySet()) {
                createJeiBuffersUnchecked(bufGroup, entry.getKey(), entry.getValue());
            }
            return bufGroup;
        }

        @SuppressWarnings("unchecked")
        private static <JB> void createJeiBuffersUnchecked(JeiBufferGroup bufGroup, BufferType<?, ?, JB, ?> bufType,
                                                           List<? extends BufferFactory<?, ?>> factories) {
            for (BufferFactory<?, JB> factory : (List<BufferFactory<?, JB>>)factories) {
                bufGroup.addBuffer(bufType, factory.createJeiBuffer());
            }
        }

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/buffer/BufferGroup.kt
package st.evening.mc.cbtweaker.buffer

import it.unimi.dsi.fastutil.objects.Object2ObjectRBTreeMap
import it.unimi.dsi.fastutil.objects.Object2ObjectSortedMap
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import st.evening.mc.cbtweaker.compat.jei.recipe.JeiBufferGroup
import st.evening.mc.cbtweaker.util.machine.MutableComponentSet
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson

typealias BufferGroups = Object2ObjectSortedMap<String, BufferGroup>

class BufferGroup {
    // iteration order must be deterministic so UI elements are constructed in the same order on the server and client
    private val bufferTable: Object2ObjectSortedMap<BufferType<*, *, *, *>, Object2ObjectSortedMap<String, *>> =
        Object2ObjectRBTreeMap(compareBy { it.id })

    @Suppress("UNCHECKED_CAST")
    fun <B> getBuffers(bufType: BufferType<B, *, *, *>): Map<String, B> =
        bufferTable[bufType] as? Map<String, B> ?: emptyMap()

    @Suppress("UNCHECKED_CAST")
    fun <B> getBuffer(bufType: BufferType<B, *, *, *>, name: String): B? = bufferTable[bufType]?.get(name) as? B

    fun <B, A> accumulateIngredients(bufType: BufferType<B, A, *, *>): A {
        val acc = bufType.createAccumulator()
        getBuffers(bufType).values.forEach {
            bufType.accumulate(acc, it)
        }
        return acc
    }

    fun <B> addBuffer(bufType: BufferType<B, *, *, *>, name: String, buffer: B) {
        val buffers = bufferTable[bufType]
        if (buffers != null) {
            @Suppress("UNCHECKED_CAST")
            (buffers as MutableMap<String, B>)[name] = buffer
        } else {
            val newBuffers = Object2ObjectRBTreeMap<String, Any>()
            newBuffers[name] = buffer
            bufferTable[bufType] = newBuffers
        }
    }

    fun forEach(visitor: Visitor) {
        bufferTable.forEach { (bufType, buffers) ->
            @Suppress("UNCHECKED_CAST")
            visitor.visit(bufType as BufferType<Any, *, *, *>, buffers as Object2ObjectSortedMap<String, Any>)
        }
    }

    interface Visitor {
        fun <B, A> visit(bufType: BufferType<B, A, *, *>, buffers: Object2ObjectSortedMap<String, B>)
    }

    class Factory {
        private val factoryTable: MutableMap<BufferType<*, *, *, *>, MutableMap<String, out BufferFactory<*, *>>> =
            mutableMapOf()

        fun <B> addFactory(bufType: BufferType<B, *, *, *>, name: String, factory: BufferFactory<B, *>) {
            val factories = factoryTable[bufType]
            if (factories != null) {
                @Suppress("UNCHECKED_CAST")
                (factories as MutableMap<String, BufferFactory<B, *>>)[name] = factory
            } else {
                factoryTable[bufType] = mutableMapOf(name to factory)
            }
        }

        context(_: JsonPath)
        fun <B> loadFactory(bufType: BufferType<B, *, *, *>, name: String, dto: TJson.Object) {
            addFactory(bufType, name, bufType.loadBufferFactory(dto))
        }

        fun createBufferGroup(world: World, pos: BlockPos, observer: BufferObserver): BufferGroup {
            val bufGroup = BufferGroup()
            factoryTable.forEach { (bufType, factories) ->
                factories.forEach { (name, factory) ->
                    @Suppress("UNCHECKED_CAST")
                    bufGroup.addBuffer(
                        bufType as BufferType<Any, *, *, *>,
                        name,
                        (factory as BufferFactory<Any, *>).createBuffer(world, pos, observer)
                    )
                }
            }
            return bufGroup
        }

        fun createJeiBufferGroup(): JeiBufferGroup {
            val bufGroup = JeiBufferGroup()
            factoryTable.forEach { (bufType, factories) ->
                factories.forEach { (name, factory) ->
                    @Suppress("UNCHECKED_CAST")
                    bufGroup.addBuffer(
                        bufType as BufferType<*, *, Any, *>,
                        name,
                        (factory as BufferFactory<*, Any>).createJeiBuffer()
                    )
                }
            }
            return bufGroup
        }
    }
}

class ComponentCollectVisitor(private val dest: MutableComponentSet) : BufferGroup.Visitor {
    override fun <B, A> visit(bufType: BufferType<B, A, *, *>, buffers: Object2ObjectSortedMap<String, B>) {
        buffers.values.forEach {
            bufType.collectComponents(dest, it)
        }
    }
}

fun Iterable<BufferGroup>.collectComponents(dest: MutableComponentSet) {
    val visitor = ComponentCollectVisitor(dest)
    forEach { it.forEach(visitor) }
}
